File: gen_resource_tag.py
import numpy as np
from datetime import timedelta

def atm_factors(n):
    return list(set((i, n//i) for i in range(2, int(n**0.5) + 1) if n % i == 0 and i < n and ((n//(i**2) < 5 and n//(i**2) > 1) or i == 3)))

def ice_factors(n):
    return list(set((i, n//i) for i in range(2, int(n**0.5) + 1) if i == 3 ))

def determine_io(in_factor):
    if in_factor % 4 == 0:
        return 4
    elif in_factor % 3 == 0:
        return 3
    elif in_factor % 2 == 0:
        return 2
    else:
        return 1

def blanks(incoming_integer):
    return ' ' * incoming_integer

class Cluster():
    def __init__(self, parent, cluster):
        self.parent = parent
        self.cluster = cluster
        if "c3" in self.cluster:
            self.cores_per_node = 32
        elif "c4" in self.cluster:
            self.cores_per_node = 36
        else:
            raise NameError

        self.ranks = self.cores_per_node * self.parent.nodes
        self.layout_cores = self.ranks / 6
        self.possible_atm_layouts = atm_factors(self.layout_cores)
        self.ice_layout = ice_factors(self.ranks)[0]

    def make_io_layouts(self):
        self.atm_io_layout = (1, determine_io(self.atm_layout[1]))
        self.ice_io_layout = (1, determine_io(self.ice_layout[0]))

    def make_layout_strings(self):
        self.atm_layout_string = '{},{}'.format(str(self.atm_layout[0]), str(self.atm_layout[1]))
        self.ice_layout_string = '{},{}'.format(str(self.ice_layout[1]), str(self.ice_layout[0]))
        self.atm_io_layout_string = '{},{}'.format(str(self.atm_io_layout[0]), str(self.atm_io_layout[1]))
        self.ice_io_layout_string = '{},{}'.format(str(self.ice_io_layout[0]), str(self.ice_io_layout[1]))

    def resolve_layouts(self):
        self.make_io_layouts()
        self.make_layout_strings()

    def write_freinclude(self):
        atm_ranks_padding = 4 - len(str(self.ranks))
        ocn_ranks_padding = 3
        blank_ranks_padding = 12
        blank_threads_padding = 11
        atm_layout_padding = 8 - len(self.atm_layout_string)
        ice_layout_padding = 8 - len(self.ice_layout_string)
        atm_io_layout_padding = 5 - len(self.atm_io_layout_string)
        ice_io_layout_padding = 5 - len(self.ice_io_layout_string)
        if 'on' in self.parent.ht:
            ht_padding = 1
        else:
            ht_padding = 0
        blank_ht_padding = 17
        string_lines = []
        string_lines.append('{}<site="{}">'.format(blanks(4), self.cluster))
        string_lines.append('{}<atm ranks="{}"{} threads="{}" layout="{}"{} io_layout="{}"{} hyperthread="{}"{} />'.format(blanks(6), str(self.ranks), blanks(atm_ranks_padding), str(self.parent.threads), self.atm_layout_string, blanks(atm_layout_padding), self.atm_io_layout_string, blanks(atm_io_layout_padding), self.parent.ht, blanks(ht_padding)))
        string_lines.append('{}<lnd {} {} layout="{}"{} io_layout="{}"{} {} />'.format(blanks(6), blanks(blank_ranks_padding), blanks(blank_threads_padding), self.atm_layout_string, blanks(atm_layout_padding), self.atm_io_layout_string, blanks(atm_io_layout_padding), blanks(blank_ht_padding)))
        string_lines.append('{}<ocn ranks="0"{} threads="0" layout="{}"{} io_layout="{}"{} hyperthread="off" />'.format(blanks(6), blanks(ocn_ranks_padding), self.ice_layout_string, blanks(ice_layout_padding), self.ice_io_layout_string, blanks(ice_io_layout_padding), blanks(blank_ht_padding)))
        string_lines.append('{}<ice {} {} layout="{}"{} io_layout="{}"{} {} />'.format(blanks(6), blanks(blank_ranks_padding), blanks(blank_threads_padding), self.ice_layout_string, blanks(ice_layout_padding), self.ice_io_layout_string, blanks(ice_io_layout_padding), blanks(blank_ht_padding)))
        string_lines.append('{}<site/>\n'.format(blanks(4)))
        return '\n'.join(string_lines)

class ResourceTag():
    def __init__(self, nodes=3, ht='off', omp='off'):
        self.nodes = nodes
        self.threads = 1
        self.ht = ht
        if 'on' in self.ht:
            self.threads *= 2
        self.omp = omp
        if 'on' in self.omp:
            self.threads *= 2
        self.name = '{}nodes_ht_{}_omp_{}'.format(str(self.nodes), self.ht, self.omp)
        self.freinclude_block = ''
        self.wallclock = self.clock_to_str()
        self.c3 = Cluster(self, "c3")
        self.c4 = Cluster(self, "c4")
        self.matched_atm_layouts = self.match_layouts() 

    def clock_to_str(self):
        clock_td = timedelta(minutes=540 // self.nodes)
        hours = clock_td.seconds//3600
        minutes = (clock_td.seconds//60) % 60
        if minutes < 3 and hours == 0:
            minutes = 3
        return '{}:{}:{}'.format(str(hours).zfill(2), str(minutes).zfill(2), str(0).zfill(2))

    def match_layouts(self):
        layout_diff = 99999
        c3idx = 0
        c4idx = 0
        for c3i, c3l in enumerate(self.c3.possible_atm_layouts):
            for c4i, c4l in enumerate(self.c4.possible_atm_layouts):
                this_diff = np.abs(c3l[0] - c4l[0]) + np.abs(c3l[1] - c4l[1])
                if this_diff < layout_diff:
                    layout_diff = this_diff
                    c3idx = c3i
                    c4idx = c4i
        self.c3.atm_layout = self.c3.possible_atm_layouts[c3idx]
        self.c3.alt_atm_layouts = list(set(self.c3.possible_atm_layouts) - set(self.c3.atm_layout))
        self.c4.atm_layout = self.c4.possible_atm_layouts[c4idx]
        self.c4.alt_atm_layouts = list(set(self.c4.possible_atm_layouts) - set(self.c4.atm_layout))

    def resolve_cluster_layouts(self):
        self.c3.resolve_layouts()
        self.c4.resolve_layouts()

    def write_prefix(self):
        string_lines = []
        string_lines.append('<freInclude name="{}">'.format(self.name))
        string_lines.append('  <resources jobWallclock="{}">\n'.format(self.wallclock))
        self.freinclude_block += '\n'.join(string_lines)
        return

    def write_suffix(self):
        string_lines = []
        string_lines.append('  </resources>')
        string_lines.append('</freInclude>')
        string_lines.append('\n')
        self.freinclude_block += '\n'.join(string_lines)
        return

    def write_all(self):
        self.write_prefix()
        self.freinclude_block += self.c3.write_freinclude()
        self.freinclude_block += self.c4.write_freinclude()
        self.write_suffix()

    def resolve_and_write(self):
        self.resolve_cluster_layouts()
        self.write_all()

def write_full(node_count_list=[3,6,9,12,24,48,96,192], ht_options=['off', 'on'], omp_options=['off', 'on'], outfile='test_file.xml'):
    for nd_count in node_count_list:
        for ht_opt in ht_options:
            for omp_opt in omp_options:
                my_rt = ResourceTag(nodes=nd_count, ht=ht_opt, omp=omp_opt)
                my_rt.resolve_and_write()
                with open(outfile, 'a+') as fh:
                    fh.write(my_rt.freinclude_block)

if __name__ == '__main__':
    write_full()

File: gen_resource_tag.jl
abstract type Layout end
abstract type BlockLine end
struct 🌩️Layout <: Layout
    layout::Tuple{Int,Int}
    🌩️Layout(layout) = new(layout)
end

struct 🧊Layout <: Layout
    layout::Tuple{Int,Int}
    🧊Layout(layout) = new(layout)
end

struct IOLayout <: Layout
    layout::Tuple{Int,Int}
    IOLayout(layout) = new(layout)
end

struct 🌩️Line <: BlockLine
    prefix::Int
    ranks::String
    🧵s::String
    🌩️::String
    🌩️_io::String
    ht::String
    function 🌩️Line(int_ranks::Int, int_🧵s::Int, layouts::Dict{String,String}, ht::String; prefix::Int = 4)
        ranks = repr(int_ranks)
        🧵s = repr(int_🧵s)
        🌩️ = layouts["🌩️"]
        🌩️_io = layouts["🌩️_io"]
        new(prefix, ranks, 🧵s, 🌩️, 🌩️_io, ht)
    end
end

struct 🌳Line <: BlockLine
    prefix::Int
    🌩️::String
    🌩️_io::String
    function 🌳Line(layouts::Dict{String,String}; prefix::Int = 4)
        🌩️ = layouts["🌩️"]
        🌩️_io = layouts["🌩️_io"]
        new(prefix, 🌩️, 🌩️_io)
    end
end

struct 🌊Line <: BlockLine
    prefix::Int
    🧊::String
    🧊_io::String
    function 🌊Line(layouts::Dict{String,String}; prefix::Int = 4)
        🧊 = layouts["🧊"]
        🧊_io = layouts["🧊_io"]
        new(prefix, 🧊, 🧊_io)
    end
end

struct 🧊Line <: BlockLine
    prefix::Int
    🧊::String
    🧊_io::String
    function 🧊Line(layouts::Dict{String,String}; prefix::Int = 4)
        🧊 = layouts["🧊"]
        🧊_io = layouts["🧊_io"]
        new(prefix, 🧊, 🧊_io)
    end
end

struct ResourceTagSetup
    nodes::Int
    ht::String
    omp::String
    ResourceTagSetup(nodes, ht, omp) = new(nodes, ht, omp)
end

struct ResourceTag
    setup::ResourceTagSetup
    🧵s::Int
    name::String
    wall🕛::String
    prefix::String
    suffix::String
    ResourceTag(setup, 🧵s, name, wall🕛, prefix, suffix) = new(setup, 🧵s, name, wall🕛, prefix, suffix)
end

struct Cluster
    ρ::ResourceTag
    κ::String
    cores_per_node::Int
    ranks::Int
    layout_cores::Int
    possible_🌩️_layouts::Array{🌩️Layout}
    🧊_layout::🧊Layout
    Cluster(ρ, κ, cores_per_node, ranks, layout_cores, possible_🌩️_layouts, 🧊_layout) = new(ρ, κ, cores_per_node, ranks, layout_cores, possible_🌩️_layouts, 🧊_layout)
end

determine_io(α::🌩️Layout) = mod(α.layout[2],4) == 0 ? 4 : mod(α.layout[2],3) == 0 ? 3 : mod(α.layout[2],2) == 0 ? 2 : 1
determine_io(ι::🧊Layout) = mod(ι.layout[1],4) == 0 ? 4 : mod(ι.layout[1],3) == 0 ? 3 : mod(ι.layout[1],2) == 0 ? 2 : 1

function calc_🧵s(σ::ResourceTagSetup)
    ht = σ.ht == "off" ? 1 : σ.ht == "on" ? 2 : throw(DomainError(σ.ht, "argument must be \"off\" or \"on\""))
    omp = σ.omp == "off" ? 1 : σ.omp == "on" ? 2 : throw(DomainError(σ.omp, "argument must be \"off\" or \"on\""))
    ht * omp
end

🌩️_factors(χ::Int) = unique([round.(Int, (i,fld(χ,i))) for i=2:floor(sqrt(χ)) if mod(χ,i) == 0 && fld(χ,i) < 25 && i < 25 && (fld(χ,i^2) < 5 || i == 3)])
🧊_factors(χ::Int) = unique([round.(Int, (i,fld(χ,i))) for i=2:floor(sqrt(χ)) if i == 3])

rt_name(σ::ResourceTagSetup) = "$(repr(σ.nodes))nodes_ht_$(σ.ht)_omp_$(σ.omp)"

function 🕛_to_str(ν::Int)
    seconds = fld(540 * 60, ν)
    hours = fld(seconds, 3600)
    minutes = hours == 0 ? max(3,mod(fld(seconds, 60), 60)) : mod(fld(seconds, 60), 60)
    "$(lpad(repr(hours),2,repr(0))):$(lpad(repr(minutes),2,repr(0))):00"
end

function make_cluster(ρ::ResourceTag, κ::String)
    cores_per_node = κ == "c3" ? 32 : κ == "c4" ? 36 : throw(DomainError(κ, "cluster not recognized, current clusters are \"c3\" and \"c4\"."))
    ranks = cores_per_node * ρ.setup.nodes
    layout_cores = div(ranks, 6)
    possible_🌩️_layouts = [🌩️Layout(🌩️_fact) for 🌩️_fact in 🌩️_factors(layout_cores)]
    🧊_layout = 🧊Layout(🧊_factors(ranks)[1])
    my_cluster = Cluster(ρ, κ, cores_per_node, ranks, layout_cores, possible_🌩️_layouts, 🧊_layout)
end

make_io_layout(λ::Layout) = IOLayout((1, determine_io(λ)))

make_layout_string(λ::Layout) = "$(λ.layout[1]),$(λ.layout[2])"

function match_layouts(γ::Array{🌩️Layout,1}, ϵ::Array{🌩️Layout,1})
    layout_diff = 99999
    c3idx = 0
    c4idx = 0
    for (c3i, c3l) in enumerate(γ)
        for (c4i, c4l) in enumerate(ϵ)
            this_diff = abs(c3l.layout[1] - c4l.layout[1]) + abs(c3l.layout[2] - c4l.layout[2])
            if this_diff < layout_diff
                layout_diff = this_diff
                c3idx = c3i
                c4idx = c4i
            end
        end
    end
    α = γ[c3idx]
    α_alt = collect(delete!(Set(γ),α))
    β = ϵ[c4idx]
    β_alt = collect(delete!(Set(ϵ),β))
    return (α, α_alt, β, β_alt)
end

function make_cluster_layouts(χ::Cluster, α::🌩️Layout)
    🌩️_layout = α
    🧊_layout = χ.🧊_layout
    🌩️_io_layout = make_io_layout(🌩️_layout)
    🧊_io_layout = make_io_layout(🧊_layout)
    proper_🧊_layout = 🧊Layout(reverse(🧊_layout.layout))
    layout_strings = Dict(x => make_layout_string(y) for (x,y) in zip(["🌩️","🧊","🌩️_io","🧊_io"],
                    [🌩️_layout, proper_🧊_layout, 🌩️_io_layout, 🧊_io_layout]))
end

function padding(ranks::Int, ht_bool::String, layouts::Dict{String,String}; 🌊_ranks::Int = 4,
                 bl_ranks::Int = 14, bl_🧵s::Int = 12, bl_ht::Int = 18, 🌩️_base::Int = 5,
                 lay_base::Int = 8, io_lay_base::Int = 5)
    🌩️_ranks = 🌩️_base - length(repr(ranks))
    🌩️ = lay_base - length(layouts["🌩️"])
    🧊 = lay_base - length(layouts["🧊"])
    🌩️_io = io_lay_base - length(layouts["🌩️_io"])
    🧊_io = io_lay_base - length(layouts["🧊_io"])
    if "$ht_bool" == "on"
        ht = 2
    else
        ht = 1
    end
    return Dict("🌩️_ranks" => 🌩️_ranks, "🌊_ranks" => 🌊_ranks, "🌩️" => 🌩️, "🧊" => 🧊,
                "🌩️_io" => 🌩️_io, "🧊_io" => 🧊_io, "ht" => ht, "bl_ht" => bl_ht,
                "bl_ranks" => bl_ranks, "bl_🧵s" => bl_🧵s)
end

function add_bl(ι::String, π::Int)
    return ι * " "^π
end

function write_block_line(α::🌩️Line, pd::Dict{String,Int})
    prefix = " "^α.prefix
    rank = add_bl("<atm ranks=\"$(α.ranks)\"", pd["🌩️_ranks"])
    🧵 = add_bl("threads=\"$(α.🧵s)\"", 1)
    layout = add_bl("layout=\"$(α.🌩️)\"", pd["🌩️"])
    io_layout = add_bl("io_layout=\"$(α.🌩️_io)\"", pd["🌩️_io"])
    ht = add_bl("hyperthread=\"$(α.ht)\"", pd["ht"])
    suffix = "/>\n"
    return "$prefix" * "$rank" * "$🧵" * "$layout" * "$io_layout" * "$ht" * "$suffix"
end

function write_block_line(λ::🌳Line, pd::Dict{String,Int})
    prefix = " "^λ.prefix
    rank = add_bl("<lnd", pd["bl_ranks"])
    🧵 = add_bl("", pd["bl_🧵s"])
    layout = add_bl("layout=\"$(λ.🌩️)\"", pd["🌩️"])
    io_layout = add_bl("io_layout=\"$(λ.🌩️_io)\"", pd["🌩️_io"])
    ht = add_bl("", pd["bl_ht"])
    suffix = "/>\n"
    return "$prefix" * "$rank" * "$🧵" * "$layout" * "$io_layout" * "$ht" * "$suffix"
end

function write_block_line(Ο::🌊Line, pd::Dict{String,Int})
    prefix = " "^Ο.prefix
    rank = add_bl("<ocn ranks=\"0\"", pd["🌊_ranks"])
    🧵 = add_bl("threads=\"0\"", 1)
    layout = add_bl("layout=\"$(Ο.🧊)\"", pd["🧊"])
    io_layout = add_bl("io_layout=\"$(Ο.🧊_io)\"", pd["🧊_io"])
    ht = add_bl("hyperthread=\"off\"", 1)
    suffix = "/>\n"
    return "$prefix" * "$rank" * "$🧵" * "$layout" * "$io_layout" * "$ht" * "$suffix"
end

function write_block_line(ι::🧊Line, pd::Dict{String,Int})
    prefix = " "^ι.prefix
    rank = add_bl("<ice", pd["bl_ranks"])
    🧵 = add_bl("", pd["bl_🧵s"])
    layout = add_bl("layout=\"$(ι.🧊)\"", pd["🧊"])
    io_layout = add_bl("io_layout=\"$(ι.🧊_io)\"", pd["🧊_io"])
    ht = add_bl("", pd["bl_ht"])
    suffix = "/>\n"
    return "$prefix" * "$rank" * "$🧵" * "$layout" * "$io_layout" * "$ht" * "$suffix"
end

function write_cluster_block(χ::String, ω::String, β::Dict{String,<:BlockLine}, pd::Dict{String,Int})
    resource_l = " "^2 * "<resources site=\"$χ\" jobWallclock=\"$ω\">\n"
    🌩️_l = write_block_line(β["🌩️"], pd)
    🌳_l = write_block_line(β["🌳"], pd)
    🌊_l = write_block_line(β["🌊"], pd)
    🧊_l = write_block_line(β["🧊"], pd)
    resource_end = " "^2 * "</resources>\n"
    return "$resource_l" * "$🌩️_l" * "$🌳_l" * "$🌊_l" * "$🧊_l" * "$resource_end"
end

function make_block_lines(ρ::Int, τ::Int, λ::Dict{String, String}, η::String)
    return Dict{String,BlockLine}("🌩️" => 🌩️Line(ρ, τ, λ, η), "🌳" => 🌳Line(λ),
                                  "🌊" => 🌊Line(λ), "🧊" => 🧊Line(λ))
end

function make_resource_tag(ν::Int, η::String, ω::String)
    my_rts = ResourceTagSetup(ν, η, ω)
    🧵s = calc_🧵s(my_rts)
    name = rt_name(my_rts)
    wall🕛 = 🕛_to_str(ν)
    prefix = "<freInclude name=\"$name\">\n"
    suffix = "</freInclude>\n\n"
    my_rt = ResourceTag(my_rts, 🧵s, name, wall🕛, prefix, suffix)
    c3 = make_cluster(my_rt, "c3")
    c4 = make_cluster(my_rt, "c4")
    c3_🌩️_layout, c3_alt_🌩️_layouts, c4_🌩️_layout, c4_alt_🌩️_layouts = match_layouts(c3.possible_🌩️_layouts, c4.possible_🌩️_layouts)
    c3_layouts = make_cluster_layouts(c3, c3_🌩️_layout)
    c4_layouts = make_cluster_layouts(c4, c4_🌩️_layout)
    c3_pd = padding(c3.ranks, my_rts.ht, c3_layouts)
    c4_pd = padding(c4.ranks, my_rts.ht, c4_layouts)
    c3_Lines = make_block_lines(c3.ranks, 🧵s, c3_layouts, my_rt.setup.ht)
    c4_Lines = make_block_lines(c4.ranks, 🧵s, c4_layouts, my_rt.setup.ht)
    c3_block = write_cluster_block("ncrc3", wall🕛, c3_Lines, c3_pd)
    c4_block = write_cluster_block("ncrc4", wall🕛, c4_Lines, c4_pd)
    full_block = "$prefix" * "$c3_block" * "$c4_block" * "$suffix"
end

function write_full(;node_count_list::Array{Int,1} = [3,6,9,12,24,48,72],
                    ht_options::Array{String, 1} = ["off", "on"],
                    omp_options::Array{String, 1} = ["off", "on"],
                    outfile="julia_test_file.xml")
    for nd_count in node_count_list
        for ht_opt in ht_options
            for omp_opt in omp_options
                open(outfile, "a+") do my_f
                    write(my_f, make_resource_tag(nd_count, ht_opt, omp_opt))
                end
            end
        end
    end
end

write_full()
